In [49]:
import numpy as np
import pandas as pd

TIEMPO
    • OverTime: Indica si el empleado trabaja horas extras ("Yes" para sí o "No" para no).
    • StandardHours: Las horas estándar de trabajo.
    • TotalWorkingYears: Total de años de experiencia laboral del empleado.
    • TrainingTimesLastYear: Número de veces que el empleado recibió capacitación el año pasado.
    • YearsAtCompany: Años que el empleado ha trabajado en la empresa actual.
    • YearsInCurrentRole: Años que el empleado ha estado en su puesto actual.
    • YearsSinceLastPromotion: Años desde la última promoción del empleado.
    • YearsWithCurrManager: Años que el empleado ha estado bajo la supervisión del actual gerente.
    • AGE

In [4]:
df = pd.read_csv("../notebooks/df_limpio.csv")
df

Unnamed: 0  age  attrition business_travel  daily_rate  \
0              0   51      False             NaN       684.0   
1              1   52      False             NaN       699.0   
2              2   42      False          rarely       532.0   
3              3   47      False          rarely       359.0   
4              4   46      False             NaN      1319.0   
...          ...  ...        ...             ...         ...   
1609        1609   36       True          rarely       530.0   
1610        1610   45      False              no       805.0   
1611        1611   39      False          rarely       903.0   
1612        1612   36      False              no      1229.0   
1613        1613   46      False             NaN       566.0   

                    department  distance_from_home  education  \
0                          NaN                   6          3   
1                          NaN                   1          4   
2      Research & Development                    4          2   
3                          NaN                   2          4   
4                          NaN                   3          3   
...                        ...                 ...        ...   
1609                       NaN                   3          1   
1610                       NaN                   4          2   
1611                       NaN                  13          5   
1612                       NaN                   8          4   
1613                       NaN                   7          2   

       education_field  employee_number  ...  work_life_balance  \
0                  NaN            162.0  ...                3.0   
1        Life Sciences            259.0  ...                3.0   
2     Technical Degree            319.0  ...                NaN   
3              Medical              NaN  ...                NaN   
4     Technical Degree              NaN  ...                3.0   
...                ...              ...  ...                ...   
1609     Life Sciences            967.0  ...                3.0   
1610               NaN            972.0  ...                2.0   
1611               NaN              NaN  ...                3.0   
1612  Technical Degree            990.0  ...                3.0   
1613           Medical              NaN  ...                3.0   

     years_at_company  years_in_current_role  years_since_last_promotion  \
0                  20                    NaN                          15   
1                  33                    NaN                          11   
2                  22                    NaN                          11   
3                  20                    NaN                           5   
4                  19                    NaN                           2   
...               ...                    ...                         ...   
1609               16                    NaN                           3   
1610                9                    NaN                           0   
1611                8                    NaN                           0   
1612                7                    NaN                           0   
1613                8                    NaN                           0   

      years_with_curr_manager same_as_monthly_income  date_birth  \
0                          15                19537.0        1972   
1                           9                19999.0        1971   
2                          15                19232.0        1981   
3                           6                17169.0        1976   
4                           8                    NaN        1977   
...                       ...                    ...         ...   
1609                        7                10325.0        1987   
1610                        8                 4447.0        1978   
1611                        7                    NaN        1984   
1612                        7                    NaN      

In [7]:
columnas_time = ['over_time','standard_hours','total_workin_gyears','training_times_last_year','years_at_company','years_in_current_role','years_since_last_promotion','years_with_curr_manager','age']

In [8]:
df[columnas_time]

over_time  standard_hours  total_workin_gyears  training_times_last_year  \
0        False             NaN                  NaN                         5   
1          NaN             NaN                 34.0                         5   
2        False             NaN                 22.0                         3   
3          NaN             NaN                  NaN                         2   
4        False             NaN                  NaN                         5   
...        ...             ...                  ...                       ...   
1609       NaN             NaN                  NaN                         6   
1610       NaN             NaN                  NaN                         5   
1611     False             NaN                  9.0                         3   
1612     False             NaN                 12.0                         3   
1613       NaN             NaN                  NaN                         3   

      years_at_company  years_in_current_role  years_since_last_promotion  \
0                   20                    NaN                          15   
1                   33                    NaN                          11   
2                   22                    NaN                          11   
3                   20                    NaN                           5   
4                   19                    NaN                           2   
...                ...                    ...                         ...   
1609                16                    NaN                           3   
1610                 9                    NaN                           0   
1611                 8                    NaN                           0   
1612                 7                    NaN                           0   
1613                 8                    NaN                           0   

      years_with_curr_manager  age  
0                          15   51  
1                           9   52  
2                          15   42  
3                           6   47  
4                           8   46  
...                       ...  ...  
1609                        7   36  
1610                        8   45  
1611                        7   39  
1612                        7   36  
1613                        7   46  

[1614 rows x 9 columns]

In [20]:
#el tipo de dato esta bien
df[columnas_time].dtypes

over_time                      object
standard_hours                float64
total_workin_gyears           float64
training_times_last_year        int64
years_at_company                int64
years_in_current_role         float64
years_since_last_promotion      int64
years_with_curr_manager         int64
age                             int64
dtype: object

In [13]:
#analizar la de overtime que es la única no numérica para olvidarnos de ella
df['over_time'].value_counts()

over_time
False    682
True     256
Name: count, dtype: int64

In [19]:
#la paso a bool aunque igual esto ya estaba hecho proque no refleja el cambio
df['over_time'].astype('bool')

0       False
1        True
2       False
3        True
4       False
        ...  
1609     True
1610     True
1611    False
1612    False
1613     True
Name: over_time, Length: 1614, dtype: bool

In [23]:
#standard hour y years in current role tienen muchos nulos, miro cuantos :
nulos_standard = (df['years_in_current_role'].isnull().sum()/ df['years_in_current_role'].shape[0])*100
nulos_standard


97.89343246592317

In [24]:
df['years_in_current_role'].value_counts()

years_in_current_role
2.0     11
7.0      5
0.0      4
4.0      3
1.0      3
11.0     2
6.0      2
3.0      2
13.0     1
12.0     1
Name: count, dtype: int64

In [ ]:
#age debe ser 18 años mayor que total working years

## volver a correr

In [57]:
df_age = df[['age', 'total_workin_gyears']].T

In [69]:
for columna in df_age:
    if df_age[columna][0] - df_age[columna][1] < 18:
        print('dato erroneo')
        print(f'la edad es :{df_age[columna][0]}')
        print(f'el total workg years es:{df_age[columna][1]}')
    
#un trabajor comenzó a trabajar con 14 años, mirar más datos sobre él para ver si el registro en general está mal

dato erroneo
la edad es :32.0
el total workg years es:18.0


In [71]:
#buscar al trabajador ----- depende del otro
trabajador14años = df.loc[(df['age'] == 32.0) & (df['total_workin_gyears'] == 18.0)]
trabajador14años

Unnamed: 0  age  attrition business_travel  daily_rate department  \
1293        1293   32      False          rarely      1305.0        NaN   

      distance_from_home  education education_field  employee_number  ...  \
1293                  10          4             NaN            518.0  ...   

      work_life_balance years_at_company  years_in_current_role  \
1293                2.0                1                    NaN   

      years_since_last_promotion  years_with_curr_manager  \
1293                           0                        1   

     same_as_monthly_income  date_birth        salary  role_departament  \
1293                 4197.0        1986  1.000000e+09               NaN   

      remote_work  
1293          NaN  

[1 rows x 40 columns]

----

In [ ]:
#el año de nacimiento tampoco está bien

In [55]:
#total working years debe ser mayor a years at company:
df_years = df[['total_workin_gyears', 'years_at_company']].T
df_years

0     1     2     3     4     5     6     7     8     \
total_workin_gyears   NaN  34.0  22.0   NaN   NaN   NaN  28.0  20.0  22.0   
years_at_company     20.0  33.0  22.0  20.0  19.0  22.0  21.0  20.0  18.0   

                     9     ...  1604  1605  1606  1607  1608  1609  1610  \
total_workin_gyears  21.0  ...   NaN   NaN   NaN  10.0   8.0   NaN   NaN   
years_at_company     18.0  ...   7.0   7.0   7.0   9.0   8.0  16.0   9.0   

                     1611  1612  1613  
total_workin_gyears   9.0  12.0   NaN  
years_at_company      8.0   7.0   8.0  

[2 rows x 1614 columns]

In [56]:
for columna in df_years:
    if {df_years[columna][0]} < {df_years[columna][1]}:
        print('dato erroneo')
        print('----------')
        
#todo en orden

In [72]:
#years_at_company debe ser mayor que years_in_current_role
df_role = df[['total_workin_gyears', 'years_in_current_role']].T
df_role

0     1     2     3     4     5     6     7     8     \
total_workin_gyears     NaN  34.0  22.0   NaN   NaN   NaN  28.0  20.0  22.0   
years_in_current_role   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

                       9     ...  1604  1605  1606  1607  1608  1609  1610  \
total_workin_gyears    21.0  ...   NaN   NaN   NaN  10.0   8.0   NaN   NaN   
years_in_current_role   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

                       1611  1612  1613  
total_workin_gyears     9.0  12.0   NaN  
years_in_current_role   NaN   NaN   NaN  

[2 rows x 1614 columns]

In [75]:
for columna in df_role:
    if {df_role[columna][0]} < {df_role[columna][1]}:
        print('dato erroneo')
        print('----------')
    
#todo en orden

In [73]:
#years_at_company debe ser mayor que years_since_last_promotion
df_promotion = df[['total_workin_gyears', 'years_since_last_promotion']].T
df_promotion

0     1     2     3     4     5     6     7     \
total_workin_gyears          NaN  34.0  22.0   NaN   NaN   NaN  28.0  20.0   
years_since_last_promotion  15.0  11.0  11.0   5.0   2.0   4.0   7.0  11.0   

                            8     9     ...  1604  1605  1606  1607  1608  \
total_workin_gyears         22.0  21.0  ...   NaN   NaN   NaN  10.0   8.0   
years_since_last_promotion  11.0   0.0  ...   1.0   5.0   0.0   8.0   6.0   

                            1609  1610  1611  1612  1613  
total_workin_gyears          NaN   NaN   9.0  12.0   NaN  
years_since_last_promotion   3.0   0.0   0.0   0.0   0.0  

[2 rows x 1614 columns]

In [76]:
for columna in df_promotion:
    if {df_promotion[columna][0]} < {df_promotion[columna][1]}:
        print('dato erroneo')
        print('----------')
    
#todo en orden

In [74]:
#years_at_company debe ser mayor que years_with_curr_manager
df_manager = df[['total_workin_gyears', 'years_with_curr_manager']].T
df_manager

0     1     2     3     4     5     6     7     8     \
total_workin_gyears       NaN  34.0  22.0   NaN   NaN   NaN  28.0  20.0  22.0   
years_with_curr_manager  15.0   9.0  15.0   6.0   8.0   7.0   9.0   6.0   8.0   

                         9     ...  1604  1605  1606  1607  1608  1609  1610  \
total_workin_gyears      21.0  ...   NaN   NaN   NaN  10.0   8.0   NaN   NaN   
years_with_curr_manager  11.0  ...   0.0   2.0   7.0   5.0   3.0   7.0   8.0   

                         1611  1612  1613  
total_workin_gyears       9.0  12.0   NaN  
years_with_curr_manager   7.0   7.0   7.0  

[2 rows x 1614 columns]

In [77]:
for columna in df_manager:
    if {df_manager[columna][0]} < {df_manager[columna][1]}:
        print('dato erroneo')
        print('----------')
    
#todo en orden

NOTAS:
- 'standard_hours' : tiene muchos nulos (74%) y el resto tienen todos el mismo dato : 80.0
- 'years_in_current_role' : tiene muchos nulos (97.9%) 

SOLUCIONAR EN EL JUPITER COMÚN:
- arreglar nombre de columna total working years (TIENE LA B ARRABAJA MAL PUESTA)
- el trabajador 1293 tiene varios fallos (la edad no concuerda con el año de nacimiento y según los años de experienci laboral empezó a trabajar con 14 años teniendo en 2023 32) pinta mal así que directamente borraría el registyro o los datos asociados a la edad/años trabajados/año nacimiento